library/std/src/sys/unix/fs.rs

```diff
@@ -751,20 +759,56 @@ impl OpenOptions {
 
 impl File {
     pub fn open(path: &Path, opts: &OpenOptions) -> io::Result<File> {
-        let path = cstr(path)?;
-        File::open_c(&path, opts)
+        let os_path = cstr(path)?;
+        let result = File::open_c(None, &os_path, opts);
+
+        #[cfg(any(target_os = "linux", target_os = "android"))]
+        let result = {
+            use crate::io::ErrorKind;
+            match result {
+                Ok(file) => Ok(file),
+                Err(e) if e.kind() == ErrorKind::FilenameTooLong => open_deep(path, opts),
+                Err(e) => Err(e),
+            }
+        };
+
+        result
     }
 
-    pub fn open_c(path: &CStr, opts: &OpenOptions) -> io::Result<File> {
+    pub fn open_c(
+        dirfd: Option<BorrowedFd<'_>>,
+        path: &CStr,
+        opts: &OpenOptions,
+    ) -> io::Result<File> {
         let flags = libc::O_CLOEXEC
             | opts.get_access_mode()?
             | opts.get_creation_mode()?
             | (opts.custom_flags as c_int & !libc::O_ACCMODE);
-        // The third argument of `open64` is documented to have type `mode_t`. On
-        // some platforms (like macOS, where `open64` is actually `open`), `mode_t` is `u16`.
-        // However, since this is a variadic function, C integer promotion rules mean that on
-        // the ABI level, this still gets passed as `c_int` (aka `u32` on Unix platforms).
-        let fd = cvt_r(|| unsafe { open64(path.as_ptr(), flags, opts.mode as c_int) })?;
+        let fd = match dirfd {
+            None => {
+                // The third argument of `open64` is documented to have type `mode_t`. On
+                // some platforms (like macOS, where `open64` is actually `open`), `mode_t` is `u16`.
+                // However, since this is a variadic function, C integer promotion rules mean that on
+                // the ABI level, this still gets passed as `c_int` (aka `u32` on Unix platforms).
+                cvt_r(|| unsafe { open64(path.as_ptr(), flags, opts.mode as c_int) })?
+            }
+            Some(dirfd) => {
+                cfg_if::cfg_if! {
+                    if #[cfg(any(target_os = "linux", target_os = "android"))] {
+                        use libc::openat64;
+                        cvt_r(|| unsafe {
```

Comment:
does this need to also indicate that the int promotion stuff applies to it as above?

Reply:
I'm not even passing mode `mode`, will fix 😔
And yes, for some 32bit androids that may be needed.